compiler/rustc_mir_build/src/builder/expr/into.rs

```diff
@@ -287,28 +288,46 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 let place = unpack!(block = this.as_place(block, expr));
                 let ty = place.ty(&this.local_decls, this.tcx).ty;
 
-                let success = this.cfg.start_new_block();
-                let clone_trait = this.tcx.require_lang_item(LangItem::Clone, None);
-                let clone_fn = this.tcx.associated_item_def_ids(clone_trait)[0];
-                let func = Operand::function_handle(this.tcx, clone_fn, [ty.into()], expr_span);
-                let ref_ty = Ty::new_imm_ref(this.tcx, this.tcx.lifetimes.re_erased, ty);
-                let ref_place = this.temp(ref_ty, span);
-                this.cfg.push_assign(
-                    block,
-                    source_info,
-                    ref_place,
-                    Rvalue::Ref(this.tcx.lifetimes.re_erased, BorrowKind::Shared, place),
-                );
-                this.cfg.terminate(block, source_info, TerminatorKind::Call {
-                    func,
-                    args: [Spanned { node: Operand::Move(ref_place), span: DUMMY_SP }].into(),
-                    destination,
-                    target: Some(success),
-                    unwind: UnwindAction::Unreachable,
-                    call_source: CallSource::Misc,
-                    fn_span: expr_span,
-                });
-                success.unit()
+                if this.tcx.type_is_copy_modulo_regions(this.infcx.typing_env(this.param_env), ty) {
+                    this.cfg.push_assign(
+                        block,
+                        source_info,
+                        destination,
+                        Rvalue::Use(Operand::Copy(place)),
+                    );
+                    block.unit()
+                } else if this.infcx.type_is_use_cloned_modulo_regions(this.param_env, ty) {
```

Comment:
this same logic needs to go into the `ExprUseVisitor`-- note that `consume_expr` I think already tests whether the expression is "copy" or not, so that is covered